### PR TITLE
tla+ spec for dynamodb lock

### PR DESCRIPTION
**Pull request by houqp**
# Description

Spec is model checked without pause and crash.

Adding crash and pause to the modeling requires enabling timer in the
spec (i.e. setting TIME_TICK_UNIT > 0). This will fail the model check
because lock with expiration cannot gurantee correctness without use of
fenching token.

### CODE REVIEW

**Comment by houqp** [comment]
also cc @mosyp who is the author of the dymodb lock.

The spec is probably not very readable for those who hasn't worked with TLA+ before. If you really want to read it, only read lines before `\* BEGIN TRANSLATION (chksum(pcal) = "e6415912" /\ chksum(tla) = "ffc65bb2")`. The rest are transpiled TLA+ code from the PlusCal algorithm code defined in the comment.

File: tlaplus/dynamodblock.tla [hunk 1]
```diff
@@ -0,0 +1,633 @@
...
+        /\ cached_lock_lookup_time_ = [ self \in ProcSet |-> NULL]
+        /\ cached_lock_duration_ = [ self \in ProcSet |-> NULL]
+        /\ cached_lock_version = [ self \in ProcSet |-> NULL]
+        (* Procedure CriticalSection *)
+        /\ local_count_value = [ self \in ProcSet |-> NULL]
+        /\ stack = [self \in ProcSet |-> << >>]
+        /\ pc = [self \in ProcSet |-> CASE self \in WRITERS -> "writerloop"
+                                        [] self = TIME_TICK_UNIT -> "tick"]
+
+cond_put(self) == /\ pc[self] = "cond_put"
```
Comment (rtyler) [inline tlaplus/dynamodblock.tla:255]: ohgod


This looks like Pascal and LaTeX joined forces with brainfuck

**Review by rtyler — COMMENTED** [review]
Uh, please add a README or something for this new folder :scream_cat:

File: tlaplus/dynamodblock.tla (hunk 1 above)
Comment (houqp) [reply tlaplus/dynamodblock.tla:259]: Yeah, it's heavily inspired by LaTex, one of the features it provides is rendering the spec in a pdf just like latex :(

**Comment by mosyp** [comment]
Wow, that is impressive diffs, how to run these specs?

**Comment by houqp** [comment]
@mosyp you will need to download tla+ toolbox and load the project to run it. See download link in the readme.

File: tlaplus/README.md [hunk 2]
```diff
@@ -0,0 +1,7 @@
+This folder contains TLA+ proofs for various subsystems of the project including:
+
+* Distributed Dynamodb lock
+* S3 multi-writer design
+
+To model check these specs, you need to use [TLA+
+toolbox](https://lamport.azurewebsites.net/tla/toolbox.html).
```
Comment (xianwill) [inline tlaplus/README.md:7]: Might be worth mentioning [this confusing behavior](https://github.com/tlaplus/tlaplus/issues/350) when opening a spec in the toolbox in the README.

**Review by xianwill — APPROVED** [review]

**Comment by Alexander-N** [comment]
Hi,
I landed here because I was searching for TLA+ applications and thought I'll share a few things which may be helpful:
* TLA+ specs can be checked from the command line, either [directly](https://github.com/lemmy/BlockingQueue) or with [other](https://github.com/pmer/tla-bin) [tools](https://github.com/hwayne/tlacli).
* To make checking the spec more straightforward I would also add a cfg file, for example as in https://github.com/fpaxos/fpaxos-tlaplus.
* If you can't get used to the latex syntax, it's possible to use VSCode and configure it so that it will be hidden: https://github.com/alygin/vscode-tlaplus/wiki/Fonts

**Pull request merged**